Building an offchain::ipfs app (#9)
rs-ipfs/offchain-ipfs-manual: src/offchain-ipfs-callables.md

# Example pallet callable reference

The following callable functions are exposed via JSON-RPC by the example pallet.
These specific functions were chosen due to popularity, to give you a familiar experience.

> However, as we explain at the bottom of this page, these functions are not the only functions

Reviewer: ```suggestion
However, as we explain at the bottom of this page, these functions are not the only ones
```

available in the native runtime, and therefore do not represent the full extent of functionality
available to you.

Regardless, we still detail the example template functions here as a helpful reference. If you have
feedback about these existing functions, or would like to request new functions, please open an issue at
[rs-ipfs/substrate].

[rs-ipfs/substrate]: https://github.com/rs-ipfs/substrate

## Callables

What follows is a list of callables, their frequency in terms of block creation, their weights,

> their arguments and return values. As explained in the previous section on the example pallet,

Reviewer: ```suggestion
their arguments and return values. As explained in the section on the example pallet,
```

weights are essentially picosecond representations of a time limit for a given transaction,
and can be loosely correlated to transaction fees.

> **Rust's `snake_case` is used here.** However, as you'll see in the upcoming polkadot.js example,
JavaScript will use `camelCase` for function and variable names. Generally, the exposed JSON-RPC
functions will adhere to the conventions of the programming language that the client code is
written in.

Also, the weights here and block frequencies here are chosen rather arbitrarily without tokenomics

> in mind. You will need to tune these values in your own custom pallet.

Reviewer: ```suggestion
in mind. You will probably need to tune these values in your own custom pallet.
```

Finally, while we list "return" values for simplicity's sake, the template pallet does not actually
return any values from the RPC calls. What this really means in the context of Substrate is that the
values will be _eventually_ emitted returned in the runtime logs, which you'll need to monitor.

### `ipfs_add_bytes`

> Adds the given bytes to the IPFS repository

Reviewer: ```suggestion
Adds the given bytes to the IPFS repository.
```

**Frequency:** Every other block<br />
**Weight**: 200,000<br />

#### Arguments

- `bytes` - `Vec<u8>`, e.g. `vec![1, 2, 3, 4]` or `b"1234"`

#### Returns

A Content ID (CID) string, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

### `ipfs_cat_bytes`

> Displays the bytes behind a given CID

Reviewer: ```suggestion
Displays the bytes behind a given CID.
```

**Frequency:** Every other block<br />
**Weight**: 100,000

#### Arguments

- `cid`: A `Vec<u8>` CID, e.g. `b"QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa"`

#### Returns

The requested bytes - UTF-8 as a string, non-UTF-8 as hexidecimal string

### `ipfs_connect`

Connects the embedded node to the given Multiaddr.

**Frequency:** Every block<br />
**Weight**: 100,000

#### Arguments

- `multiaddr`: A string, e.g. `/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ`

#### Returns

Nothing, or an error.

### `ipfs_disconnect`

Disconnects from the given Multiaddr.

**Frequency:** Every block<br />
**Weight**: 500,000

#### Arguments

- `multiaddr`: A string, e.g. `/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ`

#### Returns

Nothing, or an error.

### `ipfs_dht_findpeer`

Performs a search for the addresses associated with the provided PeerId.

**Frequency:** Every block<br />
**Weight**: 100,000

#### Arguments

- `peerID`: A `PeerId` string, e.g. `QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ`

#### Returns

> An "opaque" multiaddr, i.e. one without a `PeerId`, such as `/ip4/104.131.131.82/tcp/4001`.

Reviewer: ```suggestion
A multiaddr, such as `/ip4/104.131.131.82/tcp/4001`.
```
`OpaqueMultiaddr` doesn't necessarily mean there is no `PeerId` - it was just an alias already exposed by substrate

### `ipfs_dht_findproviders`

> Search for PeerIds known to be providing the given Cid. You must connected to at least one peer!

Reviewer: ```suggestion
Search for PeerIds known to be providing the given Cid. You must have been connected to at least one peer!
```

**Frequency:** Every block<br />
**Weight**: 100,000

#### Arguments

- `cid`: The CID string of your desired content, e.g. `QmY7Yh4UquoXHLPFo2XbhXkhBvFoPwmQUSa92pxnxjQuPU`

#### Returns

- `peerID`: An array `PeerId` string, e.g. `[QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ]`

### `ipfs_insert_pin`

Non-recursively pins a block with the specified Cid, protecting it from removal.

**Frequency:** Every other block<br />
**Weight**: 100,000

#### Arguments

- `cid`: The CID string data you want to pin, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

#### Returns

Nothing, or an error.

### `ipfs_remove_pin`

Removes a pin from a block, so that it is no longer persistent and can be removed.

**Frequency:** Every other block<br />
**Weight**: 100,000

#### Arguments

- `cid`: The CID string data you want to unpin, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

#### Returns

### `ipfs_remove_block`

> Removes a block from the node’s repository

Reviewer: ```suggestion
Removes a block from the node’s repository.
```

# Example pallet callable reference

The following callable functions are exposed via JSON-RPC by the example pallet.

> These specicif functions were chosen due to popularity to give you a familiar experience.

Reviewer: ```suggestion
These specific functions were chosen due to popularity, to give you a familiar experience.
```

However, as we explain at the bottom of this page, these functions are not the only functions
available in the native runtime, and therefore do not represent the full extent of functionality
available to you.

Regardless, we still detail the example template functions here as a helpful reference. If you have

> feedback about these existing functions, or want to request new functions, please open an issue at

Reviewer: ```suggestion
feedback about these existing functions, or would like to request new functions, please open an issue at
```

[rs-ipfs/substrate].

[rs-ipfs/substrate]: https://github.com/rs-ipfs/substrate

## Callables

What follows is a list of callables, their frequency in terms of block creation, their weights,
their arguments and return values. As explained in the previous section on the example pallet,
weights are essentially picosecond representations of a time limit for a given transaction,
and can be loosely correlated to transaction fees.

> **Rust's `snake_case` is used here.** However, as you'll see in the upcoming polkadot.js example,
JavaScript will use `camelCase` for function and variable names. Generally, the exposed JSON-RPC
functions will adhere to the conventions of the programming language that the client code is
written in.

Also, while we list "return" values for simplicity's sake, what this means in the context of

> Substrate is that the values will be _eventually_ emitted returned as chain events, which you

Reviewer 2: The pallet does not return these values at the moment, they can be found in log info.

Author: Updated in 3e217b5

will need to listen for.

### `ipfs_add_bytes`

Adds the given bytes to the IPFS repository

**Frequency:** Every other block<br />

> **Weight**: 200,000<br />

Reviewer: the weights were picked without much tokenomics in mind - we might want to remark on that :smiley:; the block frequencies were also arbitrary

Author: Updated in 3e217b5

#### Arguments

> - `bytes` - Byte array or byte string e.g. `vec![1, 2, 3, 4]` or `b"1234"`

Reviewer: on JS side it can probably be both, but strictly speaking, the pallet function expects a `Vec<u8>`

Author: Updated that... are _all_ the arguments just `Vec<u8>`

#### Returns

A Content ID (CID) string, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

### `ipfs_cat_bytes`

Displays the bytes behind a given CID

**Frequency:** Every other block<br />
**Weight**: 100,000

#### Arguments

> - `cid`: A CID string, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

Reviewer: again, substrate-wise this is technically a `Vec<u8>` (due to wasm limitations, so we should probably note that this is from the web UI or JS perspective)

Author: Updated, see above question

#### Returns

The requested bytes - UTF-8 as a string, non-UTF-8 as hexidecimal string

### `ipfs_connect`

Connects the embedded node to the given Multiaddr.

**Frequency:** Every block<br />
**Weight**: 100,000

#### Arguments

- `multiaddr`: A string, e.g. `/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ`

#### Returns

> ???

Author: Need input on the return values of this and a few other functions here

Reviewer: the return value of those is more or less "if it doesn't produce an error, it worked" :stuck_out_tongue:  plus the log info

### `ipfs_disconnect`

Disconnects from the given Multiaddr.

**Frequency:** Every block<br />
**Weight**: 500,000

#### Arguments

- `multiaddr`: A string, e.g. `/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ`

#### Returns

> ???

Author: Need input on the return values of this and a few other functions here

### `ipfs_dht_findpeer`

Performs a search for the addresses associated with the provided PeerId.

**Frequency:** Every block<br />
**Weight**: 100,000

#### Arguments

- `peerID`: A `PeerId` string, e.g. `QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ`

#### Returns

An "opaque" multiaddr, i.e. one without a `PeerId`, such as `/ip4/104.131.131.82/tcp/4001`.

### `ipfs_dht_findproviders`

Search for PeerIds known to be providing the given Cid. You must connected to at least one peer!

**Frequency:** Every block<br />
**Weight**: 100,000

#### Arguments

- `cid`: The CID string of your desired content, e.g. `QmY7Yh4UquoXHLPFo2XbhXkhBvFoPwmQUSa92pxnxjQuPU`

#### Returns

- `peerID`: An array `PeerId` string, e.g. `[QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ]`

### `ipfs_insert_pin`

Non-recursively pins a block with the specified Cid, protecting it from removal.

**Frequency:** Every other block<br />
**Weight**: 100,000

#### Arguments

- `cid`: The CID string data you want to pin, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

#### Returns

> ???

Author: Need input on the return values of this and a few other functions here

### `ipfs_remove_pin`

Removes a pin from a block, so that it is no longer persistent and can be removed.

**Frequency:** Every other block<br />
**Weight**: 100,000

#### Arguments

- `cid`: The CID string data you want to unpin, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

#### Returns

### `ipfs_remove_block`

Removes a block from the node’s repository

**Frequency:** Every other block<br />
**Weight**: 300,000

#### Arguments

- `cid`: The CID string data you want to unpin, e.g. `QmU1f6ngsoHvwtViihzLQPXCA8j3sagmvY9GJJDY7Ao7Aa`

#### Returns

> ???

Author: Need input on the return values of this and a few other functions here

## Functionality _not_ exposed in the example pallet

`offchain::ipfs` provides access to the full feature set of Rust IPFS, of which the callables above
are only a subset. In addition, the following APIs can possibly be exposed:

- Publishing pubsub messages
- Reporting bitswap Stats
- Identity and local addresses

> - Listing out local references

Author: Are these accurate?

Reviewer: there's plenty more that we don't expose from the IPFS node; I don't think we need to list them - these only have partial implementations available, which is in line with our note about how this is not the whole IPFS functionality or the only right way of exposing it

Author: Removed the whole section
